lib/internal/process/execution.js

```diff
@@ -238,10 +245,156 @@ function readStdin(callback) {
   });
 }
 
+/**
+ * Adds the TS message to the error stack.
+ *
+ * At the 3rd line of the stack, the message is added.
+ * @param {string} originalStack The stack to decorate
+ * @param {string} newMessage the message to add to the error stack
+ * @returns {void}
+ */
+function decorateCJSErrorWithTSMessage(originalStack, newMessage) {
+  let index;
+  for (let i = 0; i < 3; i++) {
+    index = StringPrototypeIndexOf(originalStack, '\n', index + 1);
+  }
+  return StringPrototypeSlice(originalStack, 0, index) +
+         '\n' + newMessage +
+         StringPrototypeSlice(originalStack, index);
+}
+
+/**
+ *
+ * Wrapper of evalScript
+ *
+ * This function wraps the evaluation of the source code in a try-catch block.
+ * If the source code fails to be evaluated, it will retry evaluating the source code
+ * with the TypeScript parser.
+ *
+ * If the source code fails to be evaluated with the TypeScript parser,
+ * it will rethrow the original error, adding the TypeScript error message to the stack.
+ *
+ * This way we don't change the behavior of the code, but we provide a better error message
+ * in case of a typescript error.
+ * @param {string} name The name of the file
+ * @param {string} source The source code to evaluate
+ * @param {boolean} breakFirstLine Whether to break on the first line
+ * @param {boolean} print If the result should be printed
+ * @param {boolean} shouldLoadESM If the code should be loaded as an ESM module
+ * @returns {void}
+ */
+function evalTypeScript(name, source, breakFirstLine, print, shouldLoadESM = false) {
+  try {
+    evalScript(name, source, breakFirstLine, print, shouldLoadESM);
```

Comment:
IIUC, this means if the code is typescript, it will be _executed_ twice, which can be quite surprising for users. It should be possible to split evalScript into parsing and execution and only do the try-catch for the parsing bit, but not the execution bit.

Reply:
Is it also possible with ESM?

Reply:
Did you mean the ESM syntax detection? Yes, for the detection part the try-catch it's purely done in parsing phase, the detection is completed before the code gets to execution.

Reply:
I mean if also the `evalModuleEntryPoint` suffers of the same problem

Reply:
Yes, though I think for `evalModuleEntryPoint` that wouldn't be too hard to split - `evalModuleEntryPoint` calls `loader.eval` which looks like this:

https://github.com/nodejs/node/blob/e85964610cbd3567ddeb740223d9ea29a8da7090/lib/internal/modules/esm/loader.js#L217-L229

Where the parsing (`compileSourceTextModule`) and linking (`new ModuleJob`)+ execution (`job.run()`) are pretty clear.

Reply:
Or technically, if you are referring to "evaluating modules" before this PR, the syntax detection was done here

https://github.com/nodejs/node/blob/94c327c753b386f8aef9754b18b6714396d1d718/lib/internal/process/execution.js#L86

and `evalModuleEntryPoint` is only run when it's already confirmed that the code is ESM.